### PR TITLE
feat: Bulk insert support fp16/bf16

### PR DESCRIPTION
**Pull request by cydrain**
Issue: #22837

### CODE REVIEW

File: internal/storage/insert_data.go [hunk 1]
```diff
@@ -616,7 +616,7 @@ func (data *BinaryVectorFieldData) AppendRows(rows interface{}) error {
 // AppendRows appends FLATTEN vectors to field data.
 func (data *FloatVectorFieldData) AppendRows(rows interface{}) error {
 	v, ok := rows.([]float32)
-	if !ok || len(v)%(data.Dim) != 0 {
```
Comment (bigsheeper) [inline internal/storage/insert_data.go:619]: Why do we need to remove this check?

File: internal/util/importutilv2/parquet/field_reader.go [hunk 2]
```diff
@@ -408,8 +405,8 @@ func ReadBinaryData(pcr *FieldReader, count int64) (any, error) {
 			}
 		case arrow.LIST:
 			listReader := chunk.(*array.List)
-			if !isRegularVector(listReader.Offsets(), pcr.dim, true) {
-				return nil, merr.WrapErrImportFailed("binary vector is irregular")
+			if !isVectorAligned(listReader.Offsets(), pcr.dim, dataType) {
+				return nil, merr.WrapErrImportFailed("binary vector not aligned")
```
Comment (bigsheeper) [inline internal/util/importutilv2/parquet/field_reader.go:409]: please refine the error to "xxx vector not aligned"

File: internal/util/importutilv2/parquet/util.go [hunk 3]
```diff
@@ -102,77 +93,116 @@ func CreateFieldReaders(ctx context.Context, fileReader *pqarrow.FileReader, sch
...
+	case schemapb.DataType_Float16Vector, schemapb.DataType_BFloat16Vector:
+		return arrow.ListOfField(arrow.Field{
+			Name:     "item",
+			Type:     &arrow.Uint8Type{},
+			Nullable: true,
+			Metadata: arrow.Metadata{},
+		})
 	default:
-		return false
+		panic("unsupported data type")
```
Comment (bigsheeper) [inline internal/util/importutilv2/parquet/util.go:150]: please do not panic

File: internal/util/importutilv2/parquet/util.go [hunk 4]
```diff
@@ -102,77 +93,116 @@ func CreateFieldReaders(ctx context.Context, fileReader *pqarrow.FileReader, sch
...
+	}
+}
+
+func isAutoPKField(field *schemapb.FieldSchema) bool {
```
Comment (bigsheeper) [inline internal/util/importutilv2/parquet/util.go:154]: move to package `typeutil` is better

Comment (cydrain) [reply internal/util/importutilv2/parquet/util.go:154]: OK

File: internal/util/importutilv2/parquet/util.go (hunk 3 above)
Comment (cydrain) [reply internal/util/importutilv2/parquet/util.go:150]: OK

File: internal/util/importutilv2/parquet/util.go [hunk 5]
```diff
@@ -102,77 +93,116 @@ func CreateFieldReaders(ctx context.Context, fileReader *pqarrow.FileReader, sch
...
+			Nullable: true,
+			Metadata: arrow.Metadata{},
+		})
+	}
+	return arrow.NewSchema(fields, nil)
+}
+
+func isSchemaEqual(schema *schemapb.CollectionSchema, arrSchema *arrow.Schema) error {
+	i := 0
+	for _, field := range schema.GetFields() {
```
Comment (bigsheeper) [inline internal/util/importutilv2/parquet/util.go:190]: just remove i, and `for i, field := range schema.GetFields() {`

File: internal/util/importutilv2/parquet/util.go [hunk 6]
```diff
@@ -102,77 +93,116 @@ func CreateFieldReaders(ctx context.Context, fileReader *pqarrow.FileReader, sch
...
+			continue
+		}
+		arrField := arrSchema.Field(i)
+		if arrField.Name != field.Name {
+			return merr.WrapErrImportFailed(
+				fmt.Sprintf("field name '%s' mis-match with arrow field name '%s", field.Name, arrField.Name))
+		}
+		toArrDataType := convertToArrowType(field.DataType)
+		if !isArrowDataTypeEqual(arrField.Type, toArrDataType) {
+			return merr.WrapErrImportFailed(fmt.Sprintf("field '%s' type mis-match", field.Name))
```
Comment (bigsheeper) [inline internal/util/importutilv2/parquet/util.go:201]: please refine the error, `field aaa type mis-match, expect bbb type, but got ccc`, if it's array field, bbb/ccc should be element type.

File: internal/util/importutilv2/parquet/util.go [hunk 7]
```diff
@@ -102,77 +93,116 @@ func CreateFieldReaders(ctx context.Context, fileReader *pqarrow.FileReader, sch
...
+		return arrow.ListOfField(arrow.Field{
+			Name:     "item",
+			Type:     &arrow.Float32Type{},
+			Nullable: true,
+			Metadata: arrow.Metadata{},
+		})
+	case schemapb.DataType_Float16Vector, schemapb.DataType_BFloat16Vector:
+		return arrow.ListOfField(arrow.Field{
+			Name:     "item",
+			Type:     &arrow.Uint8Type{},
```
Comment (bigsheeper) [inline internal/util/importutilv2/parquet/util.go:145]: What if array type is arrow.Binary?

**Comment by bigsheeper** [comment]
@xiaocai2333 please help to review the parquet part

File: internal/storage/insert_data.go (hunk 1 above)
Comment (cydrain) [reply internal/storage/insert_data.go:619]: it's duplicated with line 621

File: internal/util/importutilv2/parquet/field_reader.go (hunk 2 above)
Comment (cydrain) [reply internal/util/importutilv2/parquet/field_reader.go:409]: updated

File: internal/util/importutilv2/parquet/util.go (hunk 6 above)
Comment (cydrain) [reply internal/util/importutilv2/parquet/util.go:212]: updated

File: internal/util/importutilv2/numpy/util.go [hunk 8]
```diff
@@ -205,6 +205,13 @@ func validateHeader(npyReader *npy.Reader, field *schemapb.FieldSchema, dim int)
 		if shape[1] != dim {
 			return wrapDimError(shape[1], dim, field)
 		}
+	case schemapb.DataType_Float16Vector, schemapb.DataType_BFloat16Vector:
```
[the remark this answers is not in the document]
Comment (cydrain) [reply internal/util/importutilv2/numpy/util.go:208]: updated

File: internal/util/importutilv2/parquet/util.go (hunk 5 above)
Comment (cydrain) [reply internal/util/importutilv2/parquet/util.go:198]: if isAutoPKField() is true, i should not increase

File: internal/util/importutilv2/parquet/util.go [hunk 9]
```diff
@@ -105,77 +93,164 @@ func CreateFieldReaders(ctx context.Context, fileReader *pqarrow.FileReader, sch
...
 	case arrow.STRING:
-		return schemapb.DataType_VarChar, false
+		return dstType == arrow.STRING
 	case arrow.BINARY:
-		return schemapb.DataType_BinaryVector, false
+		return dstType == arrow.LIST && dst.(*arrow.ListType).Elem().ID() == arrow.UINT8
 	case arrow.LIST:
-		elementType, _ := convertArrowSchemaToDataType(field.Type.(*arrow.ListType).ElemField(), true)
-		return elementType, true
+		return dstType == arrow.LIST && isArrowDataTypeConvertible(src.(*arrow.ListType).Elem(), dst.(*arrow.ListType).Elem())
```
Comment (bigsheeper) [inline internal/util/importutilv2/parquet/util.go:145]: I think `dst.(*arrow.ListType)` may lead to panic

File: internal/util/importutilv2/parquet/util.go [hunk 10]
```diff
@@ -105,77 +93,164 @@ func CreateFieldReaders(ctx context.Context, fileReader *pqarrow.FileReader, sch
...
-		return schemapb.DataType_Int32, false
+		return isArrowArithmeticType(dstType) && dstType != arrow.INT8 && dstType != arrow.INT16
 	case arrow.INT64:
-		return schemapb.DataType_Int64, false
-	case arrow.FLOAT16:
-		if isList {
-			return schemapb.DataType_Float16Vector, false
-		}
-		return schemapb.DataType_None, false
+		return isArrowFloatingType(dstType) || dstType == arrow.INT64
```
Comment (bigsheeper) [inline internal/util/importutilv2/parquet/util.go:133]: Why is the `FloatingType` acceptable as well?

File: internal/util/importutilv2/parquet/util.go [hunk 11]
```diff
@@ -105,77 +93,164 @@ func CreateFieldReaders(ctx context.Context, fileReader *pqarrow.FileReader, sch
...
+		return dstType == arrow.UINT8
 	case arrow.INT8:
-		return schemapb.DataType_Int8, false
+		return isArrowArithmeticType(dstType)
 	case arrow.INT16:
-		return schemapb.DataType_Int16, false
+		return isArrowArithmeticType(dstType) && dstType != arrow.INT8
 	case arrow.INT32:
-		return schemapb.DataType_Int32, false
+		return isArrowArithmeticType(dstType) && dstType != arrow.INT8 && dstType != arrow.INT16
```
Comment (bigsheeper) [inline internal/util/importutilv2/parquet/util.go:131]: Correct me if I'm wrong, srcType is from the input file, dstType is from milvus schema, so when srcType is arrow.Int32, dstType can be `INT8` or `INT16` or `INT32`.

File: internal/util/importutilv2/parquet/util.go (hunk 7 above)
Comment (cydrain) [reply internal/util/importutilv2/parquet/util.go:203]: updated

File: internal/util/importutilv2/parquet/util.go [hunk 12]
```diff
@@ -105,77 +93,164 @@ func CreateFieldReaders(ctx context.Context, fileReader *pqarrow.FileReader, sch
...
+	return arrow.NewSchema(arrFields, nil), nil
+}
+
+func isSchemaEqual(schema *schemapb.CollectionSchema, arrSchema *arrow.Schema) error {
+	i := 0
+	for _, field := range schema.GetFields() {
+		if typeutil.IsAutoPKField(field) {
+			continue
+		}
+		arrField := arrSchema.Field(i)
```
Comment (xiaocai2333) [inline internal/util/importutilv2/parquet/util.go:238]: Not allowing fields to be in different order?

File: internal/util/importutilv2/parquet/util.go [hunk 13]
```diff
@@ -105,77 +93,164 @@ func CreateFieldReaders(ctx context.Context, fileReader *pqarrow.FileReader, sch
...
+			Name:     "item",
+			Type:     elemType,
+			Nullable: true,
+			Metadata: arrow.Metadata{},
+		}), nil
 	case schemapb.DataType_BinaryVector:
-		return dst == schemapb.DataType_BinaryVector
-	case schemapb.DataType_Float16Vector:
-		return dst == schemapb.DataType_Float16Vector
+		return arrow.ListOfField(arrow.Field{
```
Comment (xiaocai2333) [inline internal/util/importutilv2/parquet/util.go:187]: Not allowing arrow.Binary?

**Review by xiaocai2333 — COMMENTED** [review]
Losing some flexibility will lead to some compatibility issues. It is best to be compatible with the previous logic, or it needs to be discussed.

File: internal/util/importutilv2/parquet/util.go [hunk 14]
```diff
@@ -105,77 +93,164 @@ func CreateFieldReaders(ctx context.Context, fileReader *pqarrow.FileReader, sch
...
+	for _, field := range schema.GetFields() {
+		if typeutil.IsAutoPKField(field) {
+			continue
+		}
+		arrField := arrSchema.Field(i)
+		if arrField.Name != field.Name {
+			return merr.WrapErrImportFailed(
+				fmt.Sprintf("field name '%s' mis-match with arrow field name '%s", field.Name, arrField.Name))
+		}
+		toArrDataType, err := convertToArrowDataType(field, false)
```
Comment (xiaocai2333) [inline internal/util/importutilv2/parquet/util.go:243]: We can implement a function named `CanConverted(src *schempb.DataType, dst arrow.DataType) bool` ?

Comment (xiaocai2333) [reply internal/util/importutilv2/parquet/util.go:243]: for (schemapb.Binary_vector, arrow.Binary), (schemapb.Binary_vector, arrow.Uin8 list), both return true.

Comment (cydrain) [reply internal/util/importutilv2/parquet/util.go:243]: it's difficult to implement CanConverted(src *schempb.DataType, dst arrow.DataType) bool

Comment (cydrain) [reply internal/util/importutilv2/parquet/util.go:243]: > for (schemapb.Binary_vector, arrow.Binary), (schemapb.Binary_vector, arrow.Uin8 list), both return true.

what's the expected behavior ?

File: internal/util/importutilv2/parquet/util.go (hunk 13 above)
Comment (cydrain) [reply internal/util/importutilv2/parquet/util.go:187]: updated

File: internal/util/importutilv2/parquet/util.go (hunk 12 above)
Comment (cydrain) [reply internal/util/importutilv2/parquet/util.go]: updated

**Pull request merged**